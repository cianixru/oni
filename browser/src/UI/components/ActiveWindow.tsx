/**
 * ActiveWindow.tsx
 *
 * Helper component that is always sized and positioned around the currently
 * active window in Neovim.
 */

import * as React from "react"
import { connect } from "react-redux"

import * as Selectors from "./../Selectors"
import * as State from "./../State"

export interface IActiveWindowProps {
    pixelX: number
    pixelY: number
    pixelWidth: number
    pixelHeight: number
}

export class ActiveWindow extends React.PureComponent<IActiveWindowProps, {}> {
    public render(): JSX.Element {

        const px = (str: number): string => `${str}px`

        const style: React.CSSProperties = {
            position: "absolute",
            left: px(this.props.pixelX),
            top: px(this.props.pixelY),
            width: px(this.props.pixelWidth),
            height: px(this.props.pixelHeight),
        }

        return <div style={style}>
            {this.props.children}
        </div>
    }
}

const mapStateToProps = (state: State.IState): IActiveWindowProps => {
    const dimensions = Selectors.getActiveWindowPixelDimensions(state)
    return {
        pixelX: dimensions.x,
        pixelY: dimensions.y,
        pixelWidth: dimensions.width,
        pixelHeight: dimensions.height,
    }
}

export const ActiveWindowContainer = connect(mapStateToProps)(ActiveWindow)
