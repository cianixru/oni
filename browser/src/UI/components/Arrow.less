@import (reference) "./common.less";

.arrow {
    animation-duration: 0.3s;
    animation-delay: 0.2s;
    opacity: 0;
    animation-fill-mode: forwards;

    &.right {
        animation-name: appear-right;
    }

    &.left {
        animation-name: appear-left;
    }

    &.down {
        animation-name: appear-down;
    }

    &.up {
        animation-name: appear-up;
    }
}

@keyframes appear-down {
    from {
        transform: translateY(-4px);
        opacity: 0;
    }
    to {
        transform: translateY(0px);
        opacity: 1;
    }
}
@keyframes appear-up {
    from {
        transform: translateY(4px);
        opacity: 0;
    }
    to {
        transform: translateY(0px);
        opacity: 1;
    }
}
@keyframes appear-left {
    from {
        transform: translateX(4px);
        opacity: 0;
    }
    to {
        transform: translateX(0px);
        opacity: 1;
    }
}
@keyframes appear-right {
    from {
        transform: translateX(-4px);
        opacity: 0;
    }
    to {
        transform: translateX(0px);
        opacity: 1;
    }
}
