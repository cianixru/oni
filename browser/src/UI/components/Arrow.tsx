/**
 * Arrow.tsx
 *
 * Simple 'up' or 'down' arrow component
 */
import * as React from "react"

import styled from "styled-components"
import { withProps } from "./common"

export enum ArrowDirection {
    Up = 0,
    Down = 1,
    Left = 2,
    Right = 3,
}

export interface IArrowProps {
    size: number
    color: string
    direction: ArrowDirection
}

const transparentBorder = (props: IArrowProps) => `${props.size * 0.8}px solid transparent`
const solidBorder = (props: IArrowProps) => `${props.size}px solid ${props.color}`

function arrowCSS(props: IArrowProps): string {
    switch (props.direction) {
        case ArrowDirection.Up:
            return `
        width: 0px;
        height: 0px;
        border-left: ${transparentBorder(props)};
        border-right: ${transparentBorder(props)};
        border-bottom: ${solidBorder(props)};
        `
        case ArrowDirection.Down:
            return `
        width: 0px;
        height: 0px;
        border-left: ${transparentBorder(props)};
        border-right: ${transparentBorder(props)};
        border-top: ${solidBorder(props)};
        `
        case ArrowDirection.Left:
            return `
        width: 0px;
        height: 0px;
        border-top: ${transparentBorder(props)};
        border-right: ${solidBorder(props)};
        border-bottom: ${transparentBorder(props)};
        `
        case ArrowDirection.Right:
            return `
        width: 0px;
        height: 0px;
        border-top: ${transparentBorder(props)};
        border-left: ${solidBorder(props)};
        border-bottom: ${transparentBorder(props)};
        `
        default:
            return ``
    }
}

export const Arrow = withProps<IArrowProps>(styled.div)`
    ${props => arrowCSS(props)}
    `

export const preview = () => {
    return <Arrow size={100} color="yellow" direction={ArrowDirection.Left} />
}
