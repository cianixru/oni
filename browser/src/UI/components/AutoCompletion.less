
.autocompletion {
    background-color: rgb(40, 40, 40);
    border: 1px solid gray;
    box-shadow: 4px 4px rgba(0, 0, 0, 0.2);
    color: rgb(200, 200, 200);
    max-width: 800px;
    max-height: 300px;
    overflow-x: hidden;
    overflow-y: auto;
    animation-name: appear;
    animation-duration: 0.1s;

    .entry {
        &.selected {
            background-color: rgb(60, 60, 60);

            .main {
                opacity: 1;

                .icon {
                    border-left: 1px solid white;
                }
            }
        }

        .main {
            transition: opacity 1s;
            opacity: 0.8;
            display: flex;
            flex-direction: row;
            align-items: center;

            .icon {
                padding: 4px;
                width: 20px;
                flex: 0 0 auto;
                color: white;

                i {
                    margin-left: 4px;
                    margin-right: 4px;
                    font-size: 10px;
                }
            }
        }

        .label {
            flex: 1 0 auto;
            min-width:100px;
            margin-left: 8px;

            .highlight {
                font-weight: bold;
                font-style: italic;
                color: rgb(50, 200, 255);
            }
        }

        .detail {
            flex: 0 1 auto;
            min-width: 50px;
            text-align: right;

            text-overflow: ellipsis;
            overflow: hidden;
            white-space: nowrap;
            margin-left: 16px;

            color: rgba(100, 100, 100, 0.6);

            font-size: 11px;
        }

        .documentation {
            font-size: 10px;
            color: rgb(50, 200, 255);
        }
    }
}

@keyframes appear {
    from {opacity: 0;}
    to {opacity: 1;}
}

