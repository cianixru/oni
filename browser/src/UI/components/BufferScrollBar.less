@import (reference) "./common.less";

@scrollBarSize: 20px;

.scroll-bar-container {
    position: absolute;
    top: 0px;
    bottom: 0px;
    right: 0px;
    background-color: rgba(0, 0, 0, 0.2);
    width: @scrollBarSize;
    border-left: 1px solid rgba(255, 255, 255, 0.2);

    .scroll-window {
        position: absolute;
        width: @scrollBarSize;
        background-color: rgba(200, 200, 200, 0.2);
        border-top:1px solid rgba(255, 255, 255, 0.1);
        border-bottom:1px solid rgba(255, 255, 255, 0.1);
    }
}
