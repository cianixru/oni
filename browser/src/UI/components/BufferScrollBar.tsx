import * as React from "react"

import * as uniqBy from "lodash/uniqBy"
import styled from "styled-components"
import { bufferScrollBarSize, pixel, withProps } from "./common"

import { editorManager } from "./../../Services/EditorManager"

export interface IBufferScrollBarProps {
    windowId: number
    bufferSize: number
    height: number
    windowTopLine: number
    windowBottomLine: number
    markers: IScrollBarMarker[]
    visible: boolean
}

export interface IBufferScrollBarState {
    scrollBarTop: number
}

export interface IScrollBarMarker {
    line: number
    height: number
    color: string
}

const ScrollBarContainer = withProps<Partial<IScrollBarWindow>>(styled.div).attrs({
    style: ({ height }: Partial<IScrollBarWindow>) => ({
        height: pixel(height),
    }),
})`
    position: fixed;
    top: 0px;
    bottom: 0px;
    right: 0px;
    background-color: rgba(0, 0, 0, 0.2);
    width: ${bufferScrollBarSize};
    border-bottom: 1px solid black;
    pointer-events: auto;
`
interface IScrollBarWindow {
    height: number
    top: number
}

const ScrollBarWindow = withProps<IScrollBarWindow>(styled.div).attrs({
    style: ({ top, height }: IScrollBarWindow) => ({
        top: pixel(top),
        height: pixel(height),
    }),
})`
    position: absolute;
    width: ${bufferScrollBarSize};
    background-color: rgba(200, 200, 200, 0.2);
    border-top: 1px solid rgba(255, 255, 255, 0.1);
    border-bottom: 1px solid rgba(255, 255, 255, 0.1);
    pointer-events: none;
`
interface IMarkerElement {
    height: string
    top: number
    color: string
}

export const MarkerElement = withProps<IMarkerElement>(styled.div).attrs({
    style: ({ height, top }: IMarkerElement) => ({
        top: pixel(top),
        height,
    }),
})`
        background-color: ${p => p.color};
        width: 100%;
        position: absolute;
        pointer-events: none;
`

export class BufferScrollBar extends React.PureComponent<
    IBufferScrollBarProps,
    IBufferScrollBarState
> {
    public static defaultProps: Partial<IBufferScrollBarProps> = {
        markers: [],
    }

    public state: IBufferScrollBarState = {
        scrollBarTop: 0,
    }

    public setLine = (y: number) => {
        const lineFraction = Math.min(
            Math.max((y - this.state.scrollBarTop) / this.props.height, 0),
            1,
        )
        const newLine = Math.ceil(editorManager.activeEditor.activeBuffer.lineCount * lineFraction)
        editorManager.activeEditor.activeBuffer.setCursorPosition(newLine, 0)
    }

    public beginScroll = (e: React.MouseEvent<HTMLDivElement>) => {
        e.preventDefault()
        // offsetY is definitely on the scrollbar in the beginning of the click
        this.setState({ scrollBarTop: e.nativeEvent.clientY - e.nativeEvent.offsetY })
        this.setLine(e.nativeEvent.clientY)
        document.addEventListener("mousemove", this.trackScroll, true)
        document.addEventListener("mouseup", this.endScroll, true)
    }

    public trackScroll = (e: MouseEvent) => {
        e.preventDefault()
        this.setLine(e.clientY)
    }

    public endScroll = (e: MouseEvent) => {
        e.preventDefault()
        this.setLine(e.clientY)
        document.removeEventListener("mousemove", this.trackScroll, true)
        document.removeEventListener("mouseup", this.endScroll, true)
    }

    public calculateWindowDimensions() {
        const { windowBottomLine, windowTopLine, bufferSize, height } = this.props
        const windowHeight = (windowBottomLine - windowTopLine + 1) / bufferSize * height
        const windowTop = (windowTopLine - 1) / bufferSize * height
        return { windowHeight, windowTop }
    }

    public renderMarkers() {
        // Only show one marker per line in the scroll bar
        const uniqueMarkers = uniqBy(this.props.markers, ({ line }) => line)
        const markerElements = uniqueMarkers.map(({ line, color }) => {
            const pos = line / this.props.bufferSize * this.props.height
            const size = "2px"

            return (
                <MarkerElement
                    top={pos}
                    height={size}
                    color={color}
                    id="scrollbar-marker-element"
                    key={`${this.props.windowId}_${color}_${line}`}
                />
            )
        })
        return markerElements
    }

    public render() {
        if (!this.props.visible) {
            return null
        }

        const markerElements = this.renderMarkers()
        const { windowHeight, windowTop } = this.calculateWindowDimensions()

        return (
            <ScrollBarContainer
                key={this.props.windowId}
                height={this.props.height}
                onMouseDown={this.beginScroll}
            >
                <ScrollBarWindow top={windowTop} height={windowHeight} />
                {markerElements}
            </ScrollBarContainer>
        )
    }
}
