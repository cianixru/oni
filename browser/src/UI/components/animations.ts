import { keyframes } from "./common"

export const fadeInAndDown = keyframes`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
`
