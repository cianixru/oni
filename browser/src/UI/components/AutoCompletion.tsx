import * as React from "react"
import { connect } from "react-redux"

import * as _ from "lodash"

import { Icon } from "./../Icon"
import { IState /*, AutoCompletionInfo */ } from "./../State"
import { HighlightText } from "./HighlightText"

export interface IAutoCompletionProps {
    visible: boolean
    x: number
    y: number
    base: string
    entries: Oni.Plugin.CompletionInfo[]
    selectedIndex: number
}

require("./AutoCompletion.less") // tslint:disable-line no-var-requires

export class AutoCompletion extends React.Component<IAutoCompletionProps, void> {

    public render(): null | JSX.Element {

        if (!this.props.visible) {
            return null
        }

        const containerStyle = {
            position: "absolute",
            top: this.props.y.toString() + "px",
            left: this.props.x.toString() + "px",
        }

        if (this.props.entries.length === 0) {
            return null
        }

        if (this.props.entries.length === 1
            && this.props.entries[0].label === this.props.base) {
            return null
        }

        const firstTenEntries = _.take(this.props.entries, 10)

        const entries = firstTenEntries.map((s, i) => {

            const isSelected = i === this.props.selectedIndex

            return <AutoCompletionItem {...s} isSelected={isSelected} base={this.props.base} />
        })

        return (<div style={containerStyle} className="autocompletion">
            {entries}
        </div>)
    }
}

export interface IAutoCompletionItemProps extends Oni.Plugin.CompletionInfo {
    base: string
    isSelected: boolean
}

export class AutoCompletionItem extends React.Component<IAutoCompletionItemProps, void> {
    public render(): JSX.Element {

        let className = "entry"
        if (this.props.isSelected) {
            className += " selected"
        }

        const detailToShow = this.props.isSelected ? this.props.detail : ""
        const documentation = this.props.isSelected ? this.props.documentation : ""

        const highlightColor = this.props.highlightColor || this._getDefaultHighlightColor(this.props.kind as any) // FIXME: undefined

        const iconContainerStyle = {
            backgroundColor: highlightColor,
        }

        return <div className={className}>
            <div className="main">
                <span className="icon" style={iconContainerStyle}>
                    <AutoCompletionIcon kind={this.props.kind as any /* FIXME: undefined */} />
                </span>
                <HighlightText className="label" highlightClassName="highlight" highlightText={this.props.base} text={this.props.label} />
                <span className="detail">{detailToShow}</span>
            </div>
            <div className="documentation">{documentation}</div>
        </div>
    }

    private _getDefaultHighlightColor(_kind: string): string {
        // TODO: Extend this logic for better defaults per kind
        return "rgb(32, 232, 38)"
    }
}

export interface IAutoCompletionIconProps {
    kind: string
}

export class AutoCompletionIcon extends React.Component<IAutoCompletionIconProps, void> {

    public render(): null | JSX.Element {

        switch (this.props.kind) {
            case "let":
                return <Icon name="wrench" />
            case "interface":
                return <Icon name="plug" />
            case "alias":
                return <Icon name="id-badge" />
            case "const":
                return <Icon name="lock" />
            case "class":
                return <Icon name="cube" />
            case "type":
                return <Icon name="sitemap" />
            case "directory":
                return <Icon name="folder" />
            case "file":
            case "script":
                return <Icon name="file" />
            case "var":
            case "property":
            case "parameter":
                // Closed cube?
                return <Icon name="code" />
            case "module":
            case "external module name":
                return <Icon name="cubes" />
            case "method":
            case "function":
                return <Icon name="cog" />
            case "keyword":
                return <Icon name="key" />
            case "text":
                return <Icon name="align-justify" />
            case "warning":
            case "$warning":
                return <Icon name="exclamation-triangle" />
            default:
                return !this.props.kind ? null : <span>`?${this.props.kind}?`</span>
        }
    }
}

const mapStateToProps = (state: IState) => {
    if (!state.autoCompletion) {
        return {
            visible: false,
            x: state.cursorPixelX,
            y: state.cursorPixelY + state.fontPixelHeight,
            base: "",
            entries: [],
            selectedIndex: 0,
        }
    } else {
        const ret: IAutoCompletionProps = {
            visible: true,
            x: state.cursorPixelX,
            y: state.cursorPixelY + state.fontPixelHeight,
            base: state.autoCompletion.base,
            entries: state.autoCompletion.entries,
            selectedIndex: state.autoCompletion.selectedIndex,
        }
        return ret
    }
}

export const AutoCompletionContainer = connect(mapStateToProps)(AutoCompletion)
