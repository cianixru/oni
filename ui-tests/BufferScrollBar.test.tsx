import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import * as React from "react"

import { BufferScrollBar } from "./../browser/src/UI/components/BufferScrollBar"

describe("<BufferScrollBar />", () => {
    const markers = [
        {
            line: 4,
            height: 1,
            color: "yellow",
        },
        {
            line: 10,
            height: 1,
            color: "red",
        },
    ]
    it("should render without crashing", () => {
        const wrapper = shallow(
            <BufferScrollBar
                visible
                windowId={1}
                bufferSize={30}
                height={30}
                windowTopLine={1}
                windowBottomLine={25}
                markers={markers}
            />,
        )
        expect(wrapper.length).toBe(1)
    })

    it("should render the correct number of marker elements", () => {
        const wrapper = shallow(
            <BufferScrollBar
                visible
                windowId={1}
                bufferSize={30}
                height={30}
                windowTopLine={1}
                windowBottomLine={25}
                markers={markers}
            />,
        )
        expect(wrapper.dive().find("#scrollbar-marker-element").length).toBe(2)
    })

    it("should render only one marker if two marker elements share line", () => {
        const dupMarkers = [
            ...markers,
            {
                line: 10,
                height: 1,
                color: "yellow",
            },
        ]
        const wrapper = shallow(
            <BufferScrollBar
                visible
                windowId={1}
                bufferSize={30}
                height={30}
                windowTopLine={1}
                windowBottomLine={25}
                markers={dupMarkers}
            />,
        )
        expect(wrapper.dive().find("#scrollbar-marker-element").length).toBe(2)
    })

    it("should match last recorded snapshot", () => {
        const wrapper = shallow(
            <BufferScrollBar
                visible
                windowId={1}
                bufferSize={30}
                height={30}
                windowTopLine={1}
                windowBottomLine={25}
                markers={markers}
            />,
        )

        expect(toJson(wrapper)).toMatchSnapshot()
    })
})
