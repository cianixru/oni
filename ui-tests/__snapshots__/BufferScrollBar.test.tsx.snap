// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BufferScrollBar /> should match last recorded snapshot 1`] = `
<styled.div
  height={30}
  key="1"
  onMouseDown={[Function]}
>
  <styled.div
    height={25}
    top={0}
  />
  <styled.div
    color="yellow"
    height="2px"
    id="scrollbar-marker-element"
    key="1_yellow_4"
    top={4}
  />
  <styled.div
    color="red"
    height="2px"
    id="scrollbar-marker-element"
    key="1_red_10"
    top={10}
  />
</styled.div>
`;
